app/forms/api/profile_creation_form.rb

```diff
@@ -0,0 +1,152 @@
+module Api
+  class ProfileCreationForm
+    include ActiveModel::Model
+
+    validate :valid_jwt
+    validate :valid_user
+    validate :valid_password
+
+    attr_reader :password, :user_bundle
+    attr_reader :user_session, :service_provider
+    attr_reader :profile, :gpo_otp
+
+    def initialize(password:, jwt:, user_session:, service_provider: nil)
+      @password = password
+      @jwt = jwt
+      @user_session = user_session
+      @service_provider = service_provider
+      set_idv_session
+    end
+
+    def submit
+      @form_valid = valid?
+
+      if form_valid?
+        create_profile
+        cache_encrypted_pii
+        complete_session
+      end
+
+      response = FormResponse.new(
+        success: form_valid?,
+        errors: errors.to_hash,
+        extra: extra_attributes,
+      )
+      [response, personal_key]
+    end
+
+    private
+
+    attr_reader :jwt
+
+    def create_profile
+      profile_maker = build_profile_maker
+      profile = profile_maker.save_profile
+      @profile = profile
+      session[:pii] = profile_maker.pii_attributes
+      session[:profile_id] = profile.id
+      session[:personal_key] = profile.personal_key
+    end
+
+    def cache_encrypted_pii
+      cacher = Pii::Cacher.new(user, session)
+      cacher.save(password, profile)
+    end
+
+    def complete_session
+      complete_profile if phone_confirmed?
+      create_gpo_entry if user_bundle.gpo_address_verification?
+    end
+
+    def phone_confirmed?
+      user_bundle.vendor_phone_confirmation? && user_bundle.user_phone_confirmation?
+    end
+
+    def complete_profile
+      user.pending_profile&.activate
+      move_pii_to_user_session
+    end
+
+    def move_pii_to_user_session
+      return if session[:decrypted_pii].blank?
+      user_session[:decrypted_pii] = session.delete(:decrypted_pii)
+    end
+
+    def create_gpo_entry
+      move_pii_to_user_session
+      if session[:pii].is_a?(String)
+        session[:pii] = Pii::Attributes.new_from_json(user_session[:decrypted_pii])
+      end
+      confirmation_maker = GpoConfirmationMaker.new(
+        pii: session[:pii],
+        service_provider: service_provider,
+        profile: profile,
+      )
+      confirmation_maker.perform
+
+      @gpo_otp = confirmation_maker.otp
```

Comment:
Are we assigning this only for the specs? Could we we use something like `GpoConfirmation.last.otp` instead in the specs, to retrieve the OTP from the most recent database entry?

Reply:
That little bit of copy pasta is used in the original `ReviewController#create` - saved in the session if we allow autofilling of the otp (dev environments).
https://github.com/18F/identity-idp/blob/main/app/controllers/idv/review_controller.rb#L51-L52

We're going to have to revisit all of the GPO screens/logic when we get there so I'm happy to pull this out for now.

Reply:
See ddaad6b